dhcpv4/bsdp.go

```diff
@@ -0,0 +1,328 @@
+// +build darwin
+
+package dhcpv4
+
+// Implements Apple's netboot protocol BSDP (Boot Service Discovery Protocol).
+// Canonical implementation is defined here:
+// http://opensource.apple.com/source/bootp/bootp-198.1/Documentation/BSDP.doc
+
+import (
+	"encoding/binary"
+	"fmt"
+	"syscall"
+)
+
+// Options (occur as sub-options of DHCP option 43).
+const (
+	BSDPOptionMessageType OptionCode = iota + 1
+	BSDPOptionVersion
+	BSDPOptionServerIdentifier
+	BSDPOptionServerPriority
+	BSDPOptionReplyPort
+	BSDPOptionBootImageListPath // Not used
+	BSDPOptionDefaultBootImageID
+	BSDPOptionSelectedBootImageID
+	BSDPOptionBootImageList
+	BSDPOptionNetboot1_0Firmware
+	BSDPOptionBootImageAttributesFilterList
+)
+
+// Versions (seen so far)
+var (
+	BSDPVersion1_0 = []byte{1, 0}
+	BSDPVersion1_1 = []byte{1, 1}
+)
+
+// BSDP message types
+const (
+	BSDPMessageTypeList byte = iota + 1
+	BSDPMessageTypeSelect
+	BSDPMessageTypeFailed
+)
+
+// Boot image kinds
+const (
+	BSDPBootImageMacOS9 byte = iota
+	BSDPBootImageMacOSX
+	BSDPBootImageMacOSXServer
+	BSDPBootImageHardwareDiagnostics
+	// 0x4 - 0x7f are reserved for future use.
+)
+
+// BootImageID describes a boot image ID - whether it's an install image and
+// what kind of boot image (e.g. OS 9, macOS, hardware diagnostics)
+type BootImageID struct {
+	isInstall bool
+	imageKind byte
+	index     uint16
+}
+
+// toBytes serializes a BootImageID to network-order bytes.
+func (b BootImageID) toBytes() (bytes []byte) {
```

Comment:
any reason not to expose this method as public?

Reply:
Not necessarily, I just didn't think it'd need to be public given that all the serialization is handled by the exchange functions. I can make it public though to make it a bit more flexible.

Reply:
yeah, it'd be nice to make it publi public. The `BootImageID` object itself is public so anything that makes it easy to operate with it (e.g. serialization and deserialization) sounds like a good candidate for being public. The end goal for this library is to be reusable, so if someone wants to just use the BSDP subpackage but not all the rest, with public methods they can